tfsdk/data.go

```diff
@@ -0,0 +1,131 @@
+
+func (r ReadOnlyData) Get(ctx context.Context, target interface{}, options ...GetDataOption) diag.Diagnostics {
+	var reflectOpts reflect.Options
+	/*for _, option := range options {
+		// TODO: apply to reflectOpts
+	}*/
+	return reflect.Into(ctx, r.Values, target, reflectOpts)
+}
+
+func (r ReadOnlyData) GetAttribute(ctx context.Context, path *tftypes.AttributePath, target interface{}, options ...GetAttributeDataOption) diag.Diagnostics {
+	attrValue, diags := getAttributeValue(ctx, r.Values, path)
+	if diags.HasError() {
+		return diags
+	}
+
+	// TODO: handle attrValue == nil to avoid a panic
+
+	valueAsDiags := ValueAs(ctx, attrValue, target)
+
+	// ValueAs does not have path information for its Diagnostics
+	for pos, valueAsDiag := range valueAsDiags {
+		valueAsDiags[pos] = diag.WithPath(path, valueAsDiag)
+	}
+
+	diags.Append(valueAsDiags...)
+	return diags
+}
+
+func getAttributeValue(ctx context.Context, data types.Object, path *tftypes.AttributePath) (attr.Value, diag.Diagnostics) {
+	// TODO: need to walk into data to retrieve whatever's at path
+	return nil, nil
+}
+
+type Data struct {
+	ReadOnlyData
+}
+
+type SetDataOption struct{} // TODO: specify options for calling Set
+
+type SetAttributeDataOption struct{} // TODO: specify options for calling SetAttribute
+
+func (d *Data) SetAttribute(ctx context.Context, path *tftypes.AttributePath, val interface{}, options ...SetAttributeDataOption) diag.Diagnostics {
+	attrValue, diags := getAttributeValue(ctx, d.Values, path)
+	if diags.HasError() {
+		return diags
+	}
+
+	newVal, newValDiags := reflect.FromValue(ctx, attrValue.Type(ctx), val, path)
+	diags.Append(newValDiags...)
+	if diags.HasError() {
+		return diags
+	}
+
+	if attrTypeWithValidate, ok := attrValue.Type(ctx).(attr.TypeWithValidate); ok {
+		tfVal, err := newVal.ToTerraformValue(ctx)
+		if err != nil {
+			err = fmt.Errorf("error running ToTerraformValue on new state value: %w", err)
+			diags.AddAttributeError(
+				path,
+				"State Write Error",
+				"An unexpected error was encountered trying to write an attribute to the state. This is always an error in the provider. Please report the following to the provider developer:\n\n"+err.Error(),
+			)
+			return diags
+		}
+		diags.Append(attrTypeWithValidate.Validate(ctx, tfVal, path)...)
+
+		if diags.HasError() {
+			return diags
+		}
+	}
+
+	res, ds := setAttributeValue(ctx, path, d.Values, attrValue)
+	diags.Append(ds...)
+	if diags.HasError() {
+		return diags
+	}
+	d.ReadOnlyData = ReadOnlyData{Values: res}
+
+	return diags
+}
+
+func setAttributeValue(ctx context.Context, path *tftypes.AttributePath, object types.Object, val attr.Value) (types.Object, diag.Diagnostics) {
+	// TODO: update object with val at path
+	return types.Object{}, nil
+}
+
+func (d *Data) Set(ctx context.Context, val interface{}, options ...SetDataOption) diag.Diagnostics {
+	return nil
+}
+
+func (d *Data) RemoveResource(ctx context.Context) {
```

Comment:
Drive-by note: In principle, these new "data" handling types feel appropriate to centralize the data logic of the Config/Plan/State handling. However, I'm not sure that we should necessarily get away from having `Config`, `Plan`, and `State` types. This `RemoveResource` method is a good example of something that should only be implemented with `State` and not `Plan`.

I'm wondering if `Config`, `Plan`, and `State` should still exist with encapsulating the underlying data types, e.g. briefly

```go
type Config struct {
  ReadOnlyData
}

type Plan struct {
  *Data
}

type State struct {
  *Data
}

func (s State) RemoveResource(ctx context.Context) {/* ... */}
```

I'm also wondering if this particular change could be done outside the larger refactoring.